import os
from scripts.resources import Resources
include: "scripts/general_functions"
from snakemake.utils import min_version

report: "report/workflow.rst"

# Set minimum snakemake version
min_version("7.25.0")

# Load config file
configfile: "config/config.yml"

# Load gneome resources to be used in rules
resources = Resources(config["genome"], config["gencode_genome_build"])

# Get sample names
SAMPLES = utils.import_samples()

# Import rules
include: "rules/fastqc.smk"
include: "rules/trimming.smk"
include: "rules/resources.smk"
include: "rules/mapping.smk"
include: "rules/deseq2.smk"
include: "rules/plotting.smk"

# Target rule
rule all:
    input: 
        "results/plots/mapping_rates.pdf",
        "results/plots/pca.pdf",
        "results/plots/sample_distance.pdf",
        "results/plots/volcano/",
        "results/qc/multiqc/multiqc.html",

# Rules to be run on login node (very small jobs)
localrules: all, get_genome_fasta, get_trx_fasta, get_gtf

# Save snakemake terminal output to log file
snake_log = "logs/snakemake/snakemake.log"
os.makedirs("logs/snakemake", exist_ok=True)

onsuccess: 
    shell("cp -v {log} {snake_log}")

    # Compress genome files
    shell(f"pigz {resources.gencode_fasta} {resources.gencode_trx_fasta} {resources.gencode_gtf}")
    print("Analysis finished successfully!")

onerror:
    shell("cp -v {log} {snake_log}")
    print(f"Analysis (partly) failed...\nCheck {snake_log} for details")


