import os
include: "scripts/general_functions.smk"
from scripts.resources import Resources
from snakemake.utils import min_version

report: "report/workflow.rst"

# Set minimum snakemake version
min_version("7.24.0")

# Load config file
configfile: "config/config.yml"

# Load genome resources to be used in rules
resources = Resources(config["genome"], config["gencode_genome_build"])

# Get wildcard values
SAMPLES = samples()
COMPARISONS = comparisons()

# Import rules
include: "rules/fastqc.smk"
include: "rules/trimming.smk"
include: "rules/resources.smk"
include: "rules/mapping.smk"
include: "rules/deseq2.smk"
include: "rules/plotting.smk"

# Target rule
rule all:
    input: 
        targets()

# Rules to be run on login node (very small jobs)
localrules: all, get_genome_fasta, get_trx_fasta, get_gtf

# Save snakemake terminal output to log file
snake_log = "logs/snakemake/snakemake.log"
os.makedirs("logs/snakemake", exist_ok=True)

onsuccess: 
    shell("cp -v {log} {snake_log}")
    print("Analysis finished successfully!")

onerror:
    shell("cp -v {log} {snake_log}")
    print(f"Analysis failed...\nCheck {snake_log} for details")
